Feature 29/autogreet new mentors

jjhampton opened the pull request:
# Description of changes
<!-- What does this PR change and why -->
Adds a the class `member_joined_channel.rb` to respond to the `member_joined_channel` [event ](https://api.slack.com/events/member_joined_channel)from the Slack Events API.  

If the channel joined is the `#mentors-internal` channel, it will send the mentor user a welcome message with:
* Greets the user and explains the purpose of the channel
* Links to the 'Into to Mentoring' document that briefly explains the Mentorship program
* Airtable links for signing up & mentorship management
* Explains that the Slackbot will post incoming mentor requests to this channel

# Note
Requesting code review on the `member_joined_channel.rb` class, see TODO in initialize method - I'm not very familiar w/ Ruby syntax and best practices/patterns.  This class could potentially be used for other channels (possibly with their own welcome messages or actions).  There may be a more elegant way to initialize the `@template` variable than an if/else statement in the initializer.  We also may want to break out different subclasses for each channel, instead of the current setup (where the `welcome_mentor` method is inside this class.

For unit tests, I more-or-less copy/pasta'd from the existing unit tests for `team_join.rb`, with some slight modifications. Not being very familiar w/ Ruby / Mocha assertions, I used the `has_value` Mocha parameter matcher to match against the user id value.  I encountered repeated 'unexpected invocation' errors when trying other parameter matchers that were not exact matches of the `user` and `text` params when calling `Operationcode::Slack::Im.expects(:new)`.

# Issue Resolved
Closes #29

hpjaj commented:
@jjhampton  - I haven't worked in this repo before.  Let me take a look around the code to get a sense of the content and flow, and then I'll give your PR a code review.

jjhampton commented:
@hpjaj No problem, it is the first time working in the repo for me as well. I'll look at your PR #88 for the README tonight as well.

hpjaj commented on lib/event/member_joined_channel.rb:
```diff
@@ -0,0 +1,42 @@
+# Handler for a Member Joined Channel event. When a user joins a Slack channel, we want to:
+# * Check what channel is joined, and take additional actions on a per- channel basis as needed
+
+class Event
```
Instead of this being a class nested in a class, we can use the module name spacing syntax, like in [team_join.rb](https://github.com/OperationCode/operationcode_bot/blob/master/lib/event/team_join.rb#L9).

So lines 6 and 7 would be collapsed into `class Event::MemberJoinedChannel < Event`

hpjaj commented on lib/event/member_joined_channel.rb:
```diff
@@ -0,0 +1,42 @@
+      log "Production mode: #{production_mode?}"
+      log "Welcoming mentor #{resolve_user_name}"
+
+      if @channel == Event::MENTORS_INTERNAL_CHANNEL
```
If we get to this point in the code, and the channel does not meet this conditional, we will have already logged twice, but will not be sending the slack message.

Perhaps instead of the conditional being here, add a guard class as the first line in this method, on line 21, that ejects from the method if this is not the desired channel.  That way we do not log twice, with no follow up message. 

For example:

```ruby
def process
  return unless @channel == Event::MENTORS_INTERNAL_CHANNEL

  @user = Operationcode::Slack::User.new(@data['event']['user'])

  log "Production mode: #{production_mode?}"
  log "Welcoming mentor #{resolve_user_name}"

  welcome_mentor!
end
```  

On second thought, you can remove this guard clause, **and** the `@channel == Event::MENTORS_INTERNAL_CHANNEL` on line 15, and move that logic into the method that will need to be created in the `operationcode_bot.rb` file.

Then in that new method, you can do something like:

```
if it is the right channel
  call the new MemberJoinedChannel#process method
end
```

jjhampton replied:
@hpjaj That sounds like a good path to follow. The recent updates you added to the README shed some light on how `operationcode_bot.rb` ties into all of this.  I'll go over my first changeset and modify some things, most likely this weekend. Thanks for the eyes-on.

hpjaj commented on lib/event/member_joined_channel.rb:
```diff
@@ -0,0 +1,42 @@
+      im.deliver
+    end
+
+    def resolve_user_name
```
It looks like there a few opportunities to DRY up the code between this class, and the `TeamJoin` class.  Perhaps create a module that includes the common functionality, and then `include` that module in both classes?

jjhampton commented:
@hpjaj I pushed up some changes response to code review:
* Updated class namespacing to match newer convention used elsewhere in app
* Added the required 'member_joined_channel' method to `operationcode_bot.rb` that will be called when the same-named Slack Events API event occurs. I also moved the conditional logic there to check what channel was joined, and if it was the internal Mentors channel, call the process method in the newly-renamed `MentorJoinedChannel` class.  
* I thought that having a separate class for that would be easier to maintain, moving forward, especially if we wanted to do different things in different channels.
* Moved some common logic shared by `MentorJoinedChannel` & `TeamJoin` classes into a `Common` module that is included in both

hpjaj commented on lib/event.rb:
```diff
@@ -1,18 +1,14 @@
   end
 
-  # Make sure we are in dev mode unless explicitly disabled
-  def production_mode?
```
Technically, since both `team_join.rb` and `mentor_joined_channel.rb` are inheriting from `Event`, this method could remain in `event.rb`.

It works having it in `common.rb`, too.  If moving forward we feel the lion's share of the common code will live in `event.rb`, and classes are going to continue to inherit from `event.rb`, perhaps we want that method to remain there?

Your call, though.

hpjaj commented on operationcode_bot.rb:
```diff
@@ -66,6 +66,16 @@ def team_join(data, token: nil)
+def member_joined_channel(data, token: nil)
+  logger.info "Member joined channel: #{data}"
+
+  if data['event']['channel'] == Event::MENTORS_INTERNAL_CHANNEL
```
I'm not clear on how we want the behavior of this `member_joined_channel` flow to be.  For example, if this conditional is **not** met, do we still want to log this `logger.info "Member joined channel: #{data}"`, and return `empty_response`?

Or do we want to do one/neither of those things, if the conditional is met?  Also, do we want to log a different message, and return some else, if the conditional is not met?

I'm not clear on the business logic behind this, to be able to provide a recommendation.  Just food for thought.

hpjaj commented on lib/common.rb:
```diff
@@ -0,0 +1,10 @@
+module Common
+    # Make sure we are in dev mode unless explicitly disabled
+    def production_mode?
+        ENV['PRODUCTION_MODE'] == 'true' ? true : false
```
Small nit, can you pls outdent these so the file has 2 spaces across the board, vs 4 spaces.

jjhampton commented:
Closing this PR -  the current Slack bot is out of commission, and we will also eventually be moving some of this bot's Slack greeting responsibilities to the [Pybot](https://github.com/OperationCode/operation_code_pybot) once it's up and running.
